Low-level grid specification

lorenzwalthert opened the pull request:
Closes #30. Adds argument `flag_grid` to `tuning_run()` to allow an explicit specification of all combinations to process, which makes it easy to remove some. From the example fo `tuning_run()`:

```r
# using the flag_grid argument, but remove combinations
# where the combined dropout rate exceeds 1
grid <- expand.grid(
  batch_size = 2^c(1:5),
  dropout1 = seq(0.2, 0.8, by = 0.1),
  dropout2 = seq(0.2, 0.8, by = 0.1)
)
grid$combined_droput <- grid$dropout1 + grid$dropout2
grid$id <- 1:nrow(grid)
grid <- grid[grid$combined_droput <= 1,]
runs <- tuning_run("mnist_mlp.R", flag_grid = grid, sample = 0.01)
```

lorenzwalthert commented:
Potentially, the id variable in the example should be a hash of the parameter values and parameter names, not just `1:n` because when the grid changes, the id values are assigned to other parameter combinations, i.e. the mapping id -> param comb is not stable. Then we'd need the `digest` package too and write a function to hash both names *and* values which is maybe going too far for this example. We could also just add a hint that the approach with the id is brittle. What do you think?

jjallaire commented on R/training_run.R:
```diff
@@ -125,32 +125,55 @@ training_run <- function(file = "train.R",
                        context = "local",
                        config = Sys.getenv("R_CONFIG_ACTIVE", unset = "default"),
                        flags = NULL,
+                       flag_grid = NULL,
```
What would you think of just overloading the `flags` argument to take a data frame as well (in which case we'd take the `flag_grid` code path)? I think that's consistent w/ the spirit of `flags` and would be one less parameter for people to grapple with in the documentation, etc.

lorenzwalthert replied:
In general, I prefer type stability for inputs. However, I see your point since the API already has many arguments. I can adapt it if you prefer it that way.

jjallaire replied:
Yeah, in this case I think it's better we trade off on the type stability. If you could make that change it would great.

jjallaire replied:
I don't think we should have the example code be so complex/involved. Just the basic calling is all that's necessary (techniques for resuming, etc. might be better off in their own section or in a vignette).

jjallaire commented on R/training_run.R:
```diff
@@ -124,26 +124,48 @@ training_run <- function(file = "train.R",
+#' 2 %in% completed_runs
+#' # continue evaluating the grid if the process was paused / interrupted etc.
+#' grid <- grid[!(grid$id %in% completed_runs), ]
+#' runs <- tuning_run("mnist_mlp.R", flag_grid = grid, sample = 0.01)
```
This still uses the `flag_grid` variable rather than `flags`

lorenzwalthert commented:
Ok, I removed the `flag_grid` argument and dropped the complicated example with resuming etc. Also, I changed the script path in the example from `"mnist_mlp.R"` to `system.file("examples/mnist_mlp/mnist_mlp.R", package = "tfruns")` so the example works independently of the working directory or project set-up. Is that ok?
